chronos/asyncfutures2.nim

```diff
@@ -413,64 +415,68 @@ proc `or`*[T, Y](fut1: Future[T], fut2: Future[Y]): Future[void] =
   fut2.callback = cb
   return retFuture
 
-# ZAH: The return type here could be a tuple
-# This will enable waiting a heterogenous collection of futures.
+# proc all*[T](futs: varargs[Future[T]]): Future[void] =
+#   ## Returns a ``Future[void]`` which completes only when all Future[T] passed
+#   ## as arguments will be completed or failed.
+#   var completedFutures = 0
+#   let totalFutures = len(futs)
+#   var retFuture = newFuture[void]("asyncdispatch.all()")
+#   for fut in futs:
+#     fut.addCallback proc (data: pointer) =
+#       inc(completedFutures)
+#       if not retFuture.finished:
+#         if completedFutures == totalFutures:
+#           retFuture.complete()
+#   if totalFutures == 0:
+#     retFuture.complete()
+#   return retFuture
+
 proc all*[T](futs: varargs[Future[T]]): auto =
-  ## Returns a future which will complete once
-  ## all futures in ``futs`` complete.
+  ## Returns a future which will complete once all futures in ``futs`` complete.
   ## If the argument is empty, the returned future completes immediately.
   ##
   ## If the awaited futures are not ``Future[void]``, the returned future
   ## will hold the values of all awaited futures in a sequence.
   ##
-  ## If the awaited futures *are* ``Future[void]``,
-  ## this proc returns ``Future[void]``.
-
-  when T is void:
-    var
-      retFuture = newFuture[void]("asyncdispatch.all")
-      completedFutures = 0
+  ## If the awaited futures *are* ``Future[void]``, this proc returns
+  ## ``Future[void]``.
+  ##
+  ## Note, that if one of the futures in ``futs`` will fail, result of ``all()``
+  ## will also be failed with error from failed future.
+  let totalFutures = len(futs)
```

Comment:
`assert(totalFutures != 0)`?

Reply:
Please use doAssert() instead of assert().